crates/ruma-common/src/push.rs

```diff
@@ -85,6 +86,92 @@ impl Ruleset {
         self.into_iter()
     }
 
+    /// Inserts a user-defined rule in the rule set.
+    ///
+    /// If a rule with the same kind and `rule_id` exists, it will be replaced.
+    ///
+    /// If `after` or `before` is set, the rule will be moved relative to the rule with the given
+    /// ID. If both are set, the rule will be moved as the next-most important rule with respect to
+    /// `before`. If neither are set, and the rule is newly inserted, it will be moved as the rule
+    /// with the higher priority of its kind.
+    ///
+    /// Returns an error if the parameters are invalid.
+    pub fn insert(
+        &mut self,
+        rule: NewPushRule,
+        after: Option<&str>,
+        before: Option<&str>,
+    ) -> Result<(), InsertPushRuleError> {
+        let rule_id = rule.rule_id();
+        if rule_id.starts_with('.') {
+            return Err(InsertPushRuleError::ServerDefaultRuleId);
+        }
+        if rule_id.contains('/') {
+            return Err(InsertPushRuleError::InvalidRuleId);
+        }
+        if rule_id.contains('\\') {
+            return Err(InsertPushRuleError::InvalidRuleId);
+        }
+        if after.filter(|s| s.starts_with('.')).is_some() {
```

Comment:
I think `.map_or(false, |s| s.starts_with('.'))` is a little more natural to read.